### DISCUSSION TITLE
Start EMQX on Raspberry PI

### DISCUSSION THREAD

**Discussion by maticmilica** [emqx/emqx#8416] [General]
Hello, I am trying to build EMQX from the source on Raspberry PI 2B+. OS is Raspbian 10, Erlang version is 24 and I checkout to the commit 68616f1be497c35335b739c7ade9f2ddd6c78265. Make is done successfully, but I cannot start the console. I receive the following message:

```
log.console_handler.enable = EMQX_LOG__CONSOLE_HANDLER__ENABLE = true
Erlang/OTP 24 [erts-12.0] [source] [smp:4:4] [ds:4:4:8] [async-threads:4]

{"Kernel pid terminated",application_controller,"{application_start_failure,kernel,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,rocksdb}}},{kernel,start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,kernel,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,rocksdb}}},{kernel,start,[normal,[]]}}})

Crash dump is being written to: log/erl_crash.dump...done

```
Here is the erl_crash.dump file. [erl_crash.dump.txt](https://github.com/emqx/emqx/files/9042233/erl_crash.dump.txt)

Do you have any idea how to resolve this problem? I am looping around it for days now... I have tried to build it with Erlang 23 (which works on Ubuntu on PC) but I always get the same error. Also, I have tried to build the master branch with the same Erlang but I get the similar message (it says kernel pid terminated, but it fails to start **quincer_nif**).

**Comment by lafirest** [comment]
As a workaround, you can:
1. make clean-all
2.export EMQX_DB__BACKEND="mnesia"
3.export BUILD_WITHOUT_QUIC=1
4.make

**Reply by maticmilica** [reply]
Nothing has changed... Please look at the comment I posted to zmstone's answer.

**Comment by zmstone** [comment]
Seems to be due to rocksdb binary incompatible.
```
make clean-all
export BUILD_RELEASE=1
export BUILD_WITHOUT_QUIC=1
make
```
This will 
1. force rebuild rocksdb dynamic lib from source code.
2. Skip QUIC dependency

**Reply by maticmilica** [reply]
Nothing has changed... 

I forgot to write that during make, it always gets to the point where rocksdb is built and then make gets stuck. I have to restart Raspberry and then start again. After that, make passes successfully, but the error above appears when I start the emqx console. I assume that the problem is with building rocksdb, but I don't know why it is not built.

This is the last output from make:
![image](https://user-images.githubusercontent.com/17751571/177327893-d2388cd4-54b4-423d-bdaa-6d2eb5e569c6.png)

**Reply by zmstone** [reply]
sorry, my bad.
first I thought the rocksdb binary built from our release pipeline was not compatible with your platform.
but now I realise it's actually not built
https://github.com/emqx/erlang-rocksdb/blob/54657d3a44a09bd4357f07ecfa54b327c1033a17/.github/workflows/release.yaml#L82-L90

as for why your build fails then continues after reboot:
I guess it's because the failed build left behind some files made the build system to believe it's already built and continued to finish up.
then it fails to boot because the dynamic lib is not there.

I'll try to add raspbian10 to the build matrix see if it helps.
please note though:
do NOT set `export BUILD_RELEASE=1`
or set `export BUILD_RELEASE=0` instead after we pre-build the binary.

**Reply by maticmilica** [reply]
Thank you. Please let me know when I can try to run build again.

**Reply by maticmilica** [reply]
One more question. I have noticed that during the rocksdb build Raspberry's RAM memory is full, and that is the reason why make gets stuck. Is it possible to cross compile rocksdb lib? Or to limit number of process? I've noticed call of "make -j4" on lot of places.

**Reply by zmstone** [reply]
HI.  @maticmilica 
It's perhaps more complicated than I thought.
Our cross-build platform simulates `armv6l`, not sure if it's compatible to yours.
I don't have a pi to test, if you are willing to take the risk, could you help to verify a bit ?
if you write to me an email (zmstone at gmail dot com), I can reply from there.

**Reply by zmstone** [reply]
> One more question. I have noticed that during the rocksdb build Raspberry's RAM memory is full, and that is the reason why make gets stuck. Is it possible to cross compile rocksdb lib? Or to limit number of process? I've noticed call of "make -j4" on lot of places.

That's probably the reason. The build script is a bit too aggressive.

**Comment by zmstone** [comment]
related: https://github.com/emqx/emqx/issues/8457
